sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala

```diff
@@ -223,151 +277,195 @@ object CollationTypeCoercion {
     val collationContextWinner = expressions.foldLeft(findCollationContext(expressions.head)) {
       case (Some(left), right) =>
         findCollationContext(right).flatMap { ctx =>
-          collationPrecedenceWinner(left, ctx)
+          mergeWinner(left, ctx)
         }
       case (None, _) => return None
     }
-
-    collationContextWinner.flatMap { cc =>
-      extractStringType(cc.dataType)
-    }
+    collationContextWinner
   }
 
   /**
    * Tries to find the collation context for the given expression.
    * If found, it will also set the [[COLLATION_CONTEXT_TAG]] on the expression,
    * so that the context can be reused later.
    */
-  private def findCollationContext(expr: Expression): Option[CollationContext] = {
+  private def findCollationContext(expr: Expression): Option[DataType] = {
```

Comment:
Should we adjust the comment above regarding the return type?

Reply:
done!